src/libraries/Microsoft.PowerFx.Core/IR/Visitors/RewritingIRVisitor.cs

```diff
@@ -0,0 +1,251 @@
+        // If anything changes, return a new dictionary with the updates. 
+        private IReadOnlyDictionary<DName, IntermediateNode> VisitDict(IReadOnlyDictionary<DName, IntermediateNode> fields, TCtx context)
+        {
+            // Dictionary order is arbitrary, so just make a copy upfront rather than enumerate multiple times. 
+            Dictionary<DName, IntermediateNode> newFields = new Dictionary<DName, IntermediateNode>();
+
+            bool rewrite = false;
+            foreach (var kv in fields)
+            {
+                var child = Materialize(kv.Value.Accept(this, context));
+
+                newFields[kv.Key] = child;
+
+                if (!ReferenceEquals(child, kv.Value))
+                {
+                    rewrite = true;
+                }
+            }
+
+            if (rewrite)
+            {
+                return newFields;
+            }
+
+            return null;
+        }
+
+        public override TRet Visit(ErrorNode node, TCtx context)
+        {
+            return Ret(node);
+        }
+
+        public override TRet Visit(LazyEvalNode node, TCtx context)
+        {
+            var newChild = Materialize(node.Child.Accept(this, context));
+
+            if (object.ReferenceEquals(newChild, node.Child))
+            {
+                return Ret(node);
+            }
+
+            return Ret(new LazyEvalNode(node.IRContext, newChild));
+        }
+
+        public override TRet Visit(CallNode node, TCtx context)
+        {
+            // Derived visitor gets first chance. 
+            // Can callback to base if it doesn't handle. 
+
+            TRet arg0 = default;
+            if (node.Args.Count > 0)
+            {
+                arg0 = node.Args[0].Accept(this, context);
+            }
+
+            return Visit(node, context, arg0);
+        }
+
+        // Return null if no change. 
+        // Else returns a new copy ofthe list with changes. 
+        private IList<IntermediateNode> VisitList(IList<IntermediateNode> list, TCtx context, TRet arg0 = default)
+        {
+            List<IntermediateNode> newArgs = null;
+
+            for (int i = 0; i < list.Count; i++)
+            {
+                var arg = list[i];
+                var ret = (i == 0 && arg0 != null) ? arg0 : arg.Accept(this, context);
+
+                var result = Materialize(ret);
+
+                if (newArgs == null && !ReferenceEquals(arg, result))
+                {
+                    newArgs = new List<IntermediateNode>(list.Count);
+
+                    // Copy previous
+                    for (int j = 0; j < i; j++)
+                    {
+                        newArgs.Add(list[j]);
+                    }
+                }
+
+                newArgs?.Add(result);
+            }
+
+            return newArgs;
+        }
+
+        // Pass in arg0 is we already called Accept on it.
+        public TRet Visit(CallNode node, TCtx context, TRet arg0)
```

Comment:
>public [](http://example.com/codeflow?start=8&length=6)

should this one be private?

Reply:
possibly protected? Derived classes need to call it.